comp/core/tagger/impl-remote/remote.go

```diff
@@ -269,65 +272,78 @@ func (t *remoteTagger) GenerateContainerIDFromOriginInfo(originInfo origindetect
 		}
 	}()
 
-	expBackoff := backoff.NewExponentialBackOff()
-	expBackoff.InitialInterval = 500 * time.Millisecond
-	expBackoff.MaxInterval = 1 * time.Second
-	expBackoff.MaxElapsedTime = 15 * time.Second
-
-	var containerID string
+	// Generate cache key
+	initPrefix := ""
+	if originInfo.ExternalData.Init {
+		initPrefix = "i/"
+	}
+	key := cache.BuildAgentKey(
+		"remoteTagger",
+		"cid",
+		initPrefix+originInfo.ExternalData.PodUID+"/"+originInfo.ExternalData.ContainerName,
+	)
 
-	err := backoff.Retry(func() error {
-		select {
-		case <-t.ctx.Done():
-			return &backoff.PermanentError{Err: errTaggerFailedGenerateContainerIDFromOriginInfo}
-		default:
-		}
+	cachedContainerID, err := cache.GetWithExpiration(key, func() (containerID string, err error) {
+		expBackoff := backoff.NewExponentialBackOff()
+		expBackoff.InitialInterval = 200 * time.Millisecond
+		expBackoff.MaxInterval = 1 * time.Second
+		expBackoff.MaxElapsedTime = 15 * time.Second
+
+		err = backoff.Retry(func() error {
+			select {
+			case <-t.ctx.Done():
+				return &backoff.PermanentError{Err: errTaggerFailedGenerateContainerIDFromOriginInfo}
+			default:
+			}
 
-		// Fetch the auth token
-		if t.token == "" {
-			var authError error
-			t.token, authError = t.options.TokenFetcher()
-			if authError != nil {
-				_ = t.log.Errorf("unable to fetch auth token, will possibly retry: %s", authError)
-				return authError
+			// Fetch the auth token
+			if t.token == "" {
+				var authError error
+				t.token, authError = t.options.TokenFetcher()
+				if authError != nil {
+					_ = t.log.Errorf("unable to fetch auth token, will possibly retry: %s", authError)
+					return authError
+				}
 			}
-		}
 
-		// Create the context with the auth token
-		t.queryCtx, t.queryCancel = context.WithCancel(
-			metadata.NewOutgoingContext(t.ctx, metadata.MD{
-				"authorization": []string{fmt.Sprintf("Bearer %s", t.token)},
-			}),
-		)
+			// Create the context with the auth token
+			t.queryCtx, t.queryCancel = context.WithCancel(
```

Comment:
I think we have a race here if several goroutine cal the function `GenerateContainerIDFromOriginInfo` 

`queryCtx` and `queryCancel` need to be local and not part of the `remoteTagger` struct

Reply:
Agreed, done 👍